File: apps/cli/src/lib/check-for-updates.ts
import { version } from "config";
import { getLastShownUpdateVersion, setLastShownUpdateVersion } from "node-api";

//
// URL of the GitHub API endpoint that returns the latest non-prerelease release.
//
const LATEST_RELEASE_URL = "https://api.github.com/repos/ashleydavis/photosphere/releases/latest";

//
// Shape of the response object returned by the GitHub releases/latest endpoint.
//
interface IGitHubReleaseResponse {
    //
    // The git tag for the release (e.g. "v1.2.3").
    //
    tag_name: string;
}

//
// Checks GitHub for the latest Photosphere release and returns the version string
// (without leading "v") when it differs from the running version AND has not
// already been notified to the user (per news.yaml's last_shown_update_version).
// Returns undefined when the running version is current, when it is a non-release
// build ("dev" or nightly), when the user has already been notified about this
// version, or when the network/parse step fails.
//
export async function checkForUpdates(): Promise<string | undefined> {
    const currentVersion: string = version;
    if (currentVersion === "dev" || currentVersion.includes("nightly")) {
        return undefined;
    }

    try {
        const response = await fetch(LATEST_RELEASE_URL);
        if (!response.ok) {
            return undefined;
        }
        const data = await response.json() as IGitHubReleaseResponse;
        if (!data.tag_name || typeof data.tag_name !== "string") {
            return undefined;
        }
        const latestVersion = data.tag_name.startsWith("v")
            ? data.tag_name.slice(1)
            : data.tag_name;
        if (latestVersion === currentVersion) {
            return undefined;
        }
        const lastShown = await getLastShownUpdateVersion();
        if (lastShown === latestVersion) {
            return undefined;
        }
        return latestVersion;
    }
    catch (error) {
        return undefined;
    }
}

//
// Records that the user has been notified about the given update version, so
// subsequent checkForUpdates() calls suppress the notification until a newer
// version ships. Persistence failures are swallowed silently.
//
export async function markUpdateAsShown(latestVersion: string): Promise<void> {
    try {
        await setLastShownUpdateVersion(latestVersion);
    }
    catch (error) {
        // Update persistence failures must never block the user.
    }
}

//
// Returns the latest release version reported by GitHub (without the leading "v"),
// or undefined when the running build is dev/nightly or the fetch/parse step fails.
// Unlike checkForUpdates(), this does NOT apply the `last_shown_update_version` dedup,
// does NOT compare to the running version, and does NOT record anything. Used by
// `psi news` to always show the latest available version when known.
//
export async function getLatestVersion(): Promise<string | undefined> {
    const currentVersion: string = version;
    if (currentVersion === "dev" || currentVersion.includes("nightly")) {
        return undefined;
    }
    try {
        const response = await fetch(LATEST_RELEASE_URL);
        if (!response.ok) {
            return undefined;
        }
        const data = await response.json() as IGitHubReleaseResponse;
        if (!data.tag_name || typeof data.tag_name !== "string") {
            return undefined;
        }
        return data.tag_name.startsWith("v")
            ? data.tag_name.slice(1)
            : data.tag_name;
    }
    catch (error) {
        return undefined;
    }
}

File: apps/cli-zig/src/lib/check-for-updates.zig
const std = @import("std");
const node_api = @import("node-api-zig");
const config = @import("config.zig");
const fetch = node_api.fetch.fetch;
const getLastShownUpdateVersion = node_api.news_state.getLastShownUpdateVersion;
const setLastShownUpdateVersion = node_api.news_state.setLastShownUpdateVersion;

//
// URL of the GitHub API endpoint that returns the latest non-prerelease release.
//
const LATEST_RELEASE_URL = "https://api.github.com/repos/ashleydavis/photosphere/releases/latest";

// Not ported: IGitHubReleaseResponse (the response is a std.json.Value here).

//
// The tag_name of the release response, or null when it is missing or not a string.
//
fn tagName(allocator: std.mem.Allocator, body: []const u8) ?[]const u8 {
    const data = std.json.parseFromSliceLeaky(std.json.Value, allocator, body, .{}) catch return null;
    if (data != .object) {
        return null;
    }
    const tag = data.object.get("tag_name") orelse return null;
    if (tag != .string or tag.string.len == 0) {
        return null;
    }
    return tag.string;
}

//
// The body of checkForUpdates inside its try block (errors become undefined).
//
fn checkForUpdatesUnsafe(allocator: std.mem.Allocator, io: std.Io, currentVersion: []const u8) !?[]const u8 {
    const response = try fetch(allocator, io, LATEST_RELEASE_URL);
    if (!response.ok) {
        return null;
    }
    const tag = tagName(allocator, response.body) orelse return null;
    const latestVersion = if (std.mem.startsWith(u8, tag, "v")) tag[1..] else tag;
    if (std.mem.eql(u8, latestVersion, currentVersion)) {
        return null;
    }
    const lastShown = try getLastShownUpdateVersion(allocator, io);
    if (lastShown != null and std.mem.eql(u8, lastShown.?, latestVersion)) {
        return null;
    }
    return latestVersion;
}

//
// Checks GitHub for the latest Photosphere release and returns the version string
// (without leading "v") when it differs from the running version AND has not
// already been notified to the user (per news.yaml's last_shown_update_version).
// Returns undefined when the running version is current, when it is a non-release
// build ("dev" or nightly), when the user has already been notified about this
// version, or when the network/parse step fails.
//
pub fn checkForUpdates(allocator: std.mem.Allocator, io: std.Io) ?[]const u8 {
    const currentVersion: []const u8 = config.version;
    if (std.mem.eql(u8, currentVersion, "dev") or std.mem.indexOf(u8, currentVersion, "nightly") != null) {
        return null;
    }

    return checkForUpdatesUnsafe(allocator, io, currentVersion) catch null;
}

//
// Records that the user has been notified about the given update version, so
// subsequent checkForUpdates() calls suppress the notification until a newer
// version ships. Persistence failures are swallowed silently.
//
pub fn markUpdateAsShown(allocator: std.mem.Allocator, io: std.Io, latestVersion: []const u8) void {
    setLastShownUpdateVersion(allocator, io, latestVersion) catch {
        // Update persistence failures must never block the user.
    };
}

// Not ported: getLatestVersion (only used by psi news).
